src/fhirtypes/common.ts

```diff
@@ -1602,3 +1604,50 @@ export function findImposeProfiles(sd: any): string[] | undefined {
     return imposeProfiles;
   }
 }
+
+// Used to check if the entity used in Canonical() references a contained resource.
+// Checks a list of validated rules at any contained path for a matching value from the
+// Canonical() keyword and returns the matching resource's id.
+export function getMatchingContainedReferenceId(
+  value: string,
+  containedResources: { pathParts: PathPart[]; assignedValue: any }[]
+) {
+  const matchingContainedResource = containedResources.find(
+    r =>
+      r.assignedValue?.url === value ||
+      r.assignedValue?.name === value ||
+      r.assignedValue?.id === value
+  );
+  if (matchingContainedResource != null) {
+    return matchingContainedResource.assignedValue.id;
+  }
```

Comment:
If the contained resource rules can be any rule starting w/ `contained`, then isn't it possible that the assigned value might be at a nested element within a contained resource? E.g.
```
* contained[0].resourceType = "Bundle"
* contained[0].entry[0].resource = MyPlanDefinitionInstance
```
Now if a canonical reference referred to `MyPlanDefinitionInstance`'s url, name, or id, I think this code would create the fragment even though `MyPlanDefinitionInstance` isn't a directly contained resource. But I don't think we want that.

I'd actually suggest that in the code where we build up the list of contained resources (in `InstanceExporter`), we use a more picky regex that only catches paths that end with the `contained` element or end with `url`/`name`/`id` directly on the contained element.

So... collect these ones:
* `contained = Foo`
* `contained[0] = Foo`
* `contained[sliceName] = Foo`
* `contained.url = "http://example.org/foo"`
* `contained[0].name = "Foo"`
* `contained[sliceName].id = "foo"`
* _etc._

But do _NOT_ collect these ones:
* `contained.status = #final`
* `contained[0].entry[0].resource = MyPlanDefinitionInstance`
* `contained[mySlice].entry.resource.id = "foo"`
* _etc._

This will prevent us from matching on resources nested in a contained resource _and_ it should make the looping faster since you're only looping through the truly relevant rules.

Reply:
[8977105](https://github.com/FHIR/sushi/pull/1405/commits/8977105549e9cdda2a0f8a6981a304ba33873e38) updates the regex and adds a test case to avoid this edge case.